### DISCUSSION TITLE
"zpool replace" mechanism?

### DISCUSSION THREAD

**Discussion by zhynwng** [openzfs/zfs#16354] [Q&A]
Dear ZFS community,

I'm currently researching OpenZFS repair code, and I'm a bit confused about the mechanism of the "zpool replace" command.

In my setup, I first created three 4GB files as virtual hard drives and then set up a 2+1 pool with a RAIDZ1 configuration. I then zeroed out the first 80MB of the first virtual hard drive, which resulted in an "invalid label" error, as shown below:

> pool: zpool
> state: DEGRADED
> status: One or more devices could not be used because the label is missing or
>           invalid.  Sufficient replicas exist for the pool to continue
>           functioning in a degraded state.
> action: Replace the device using 'zpool replace'.
> see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-4J
> scan: scrub repaired 0B in 00:00:00 with 0 errors on Tue Jul 16 04:21:14 2024
> config:
>   
>           NAME                STATE     READ WRITE CKSUM
>           zpool               DEGRADED     0     0     0
>             raidz1-0          DEGRADED     0     0     0
>               /scratch/1.img  UNAVAIL      0     0     0  corrupted data
>               /scratch/2.img  ONLINE       0     0     0
>               /scratch/3.img  ONLINE       0     0     0

I was unable to fix the "invalid label" error using the "zpool scrub" command. However, when I replaced the "failed" disk with a new one using the "zpool replace" command, the "invalid label" error disappeared. I'm wondering how the "zpool replace" command fixes the invalid label error. Does it create a new label on the new disk? How does this new label integrate into the pool? A high-level explaination would be great!

I would greatly appreciate any help!

**Comment by amotin** [comment]
`zpool replace` re-adds the disk into the pool, reconstructing the data completely using available redundancy, so there is no any surprise that it worked.  Same time it is a bit surprising to me that with 2 out of 4 labels of the vdev being erased the disk became unusable, I'd expect it to be recoverable, though I haven't tried it lately.
